src/txn/txn_ckpt.c

```diff
@@ -189,10 +211,19 @@ __checkpoint_apply(WT_SESSION_IMPL *session, const char *cfg[],
 			continue;
 		WT_WITH_DHANDLE(session, session->ckpt_handle[i],
 		    ret = (*op)(session, cfg));
-		WT_RET(ret);
+		WT_ERR(ret);
 	}
 
-	return (0);
+err:
+	/*
+	 * If we have an error somewhere in processing the handles, then
+	 * we need to mark earlier trees dirty.  Recursively call the apply
+	 * function to reset on error.
+	 */
+	if (ret != 0)
+		(void)__checkpoint_apply(session, cfg, __checkpoint_err_reset);
```

Comment:
Moving it up to `txn_checkpoint` seems messier because there are a lot of different paths to get an error that are not necessarily related to walking the array of handles.  I'm more of a fan of dealing with the error closer to where it happened.  Would something additional in the comment help your comfort level?  @keithbostic do you have an opinion on this error path?